woocommerce/woocommerce-ios
# Send receipts after failed payment to attached customer email

Show "A receipt has been sent to {email}" message after a failed payment when the order has a customer email set.

Resolves: #14163
I'll address failed payments sent via the API in the next task.



## Description


I planned to implement the whole remaining failure receipt functionality within a single PR but it already grew more complex than I anticipated.

In this PR I address the functionality of sending receipts after failed payment to attached customer and displaying "Email sent" message. Equivalent of https://github.com/woocommerce/woocommerce-ios/pull/14390 but to failed payments.

## Solution

Reminder of backend changes:
- From WooCommerce 9.5, receipts are emailed automatically to the attached customer after the order status is changed to failed
- From WooPayments 8.6, the order status is changed to failed when a payment fails,  now using the same webhooks for card_present payments as with card payments

I had to make more changes than anticipated before development:
- Only failures within the payment processor when making a payment result in an order-failed status change. Other types of failures could be local: a network connection, reader connection, sdk issues, etc... Those don't get picked up by a WooPayments webhook and don't send order receipts automatically. 
- Order can succeed only once while payment can fail multiple times. I had to figure out how to handle it or communicate with the user. The backend only sends the receipt when the order turns to failed from another type of status.

1. I expanded the card presentation failure models with two more options shown when a payment fails.

| Class                               | Description                                                                                      |
|------------------------------------|--------------------------------------------------------------------------------------------------
| `CardPresentModalErrorEmailSent` | Is shown if the failed payment has customer email set
| `CardPresentModalNonRetryableErrorEmailSent` | Is shown if the failed has customer email set and is not retryable anymore                                     |

2. Show email sent only for `CardReaderServiceError.paymentCapture(.paymentDeclinedByPaymentProcessorAPI)` errors. Other types of errors related to order preparation, card readers, and network issues don't produce failed order status. I haven't found any other way to determine whether the order turned into a failed state. Retrieving an order and checking status after payment failure is unreliable since webhook usually doesn't manage to update the order status before we make a call. I will use the same logic to display a button to send failure receipts as well.
3. When an order is in `failed` status and is retried and fails again, the new receipt doesn't get sent. To avoid this, I set the order status back to `pending` when the order payment is collected again.  

## Steps to reproduce


Before testing:
1. Enable `sendReceiptAfterPayment` in `DefaultFeatureFlagService`
2. Use my testing site https://lovely-aware-sturgeon.jurassic.ninja/ details: PdfdoF-5Qj-p2

Test cases:
- Failed Card Reader Payment: Create an order, attach a customer with an email, [set a failure amount](https://docs.stripe.com/terminal/references/testing#physical-test-cards), and pay with a card reader. Confirm 'Email sent' message is shown and the email is received.
- Failed TTP payment: Create an order, attach a customer with an email, [set a failure amount](https://docs.stripe.com/terminal/references/testing#physical-test-cards), and pay with TTP. Confirm 'Email sent' message is shown and the email is received.
- Try Collecting Again: After payment failure, try collecting again, Confirm the 'Email sent' message is shown.
- Payment-unrelated failure: Create an order, attach a customer with an email, trigger a card payment, instead of tapping a card, disable Bluetooth on the phone. Confirm no 'Email sent' message is displayed and email is not received.
- Try Collecting Later: Make 'Failed Card Reader Payment', come back to the order screen, tap 'Collect Payment' again, try to make a card reader.  Confirm 'Email sent' message is shown and the email is received.

## Testing information


The feature is still behind the feature flag. Added new unit tests.

## Screenshots


<img src="https://github.com/user-attachments/assets/458eed78-0a10-4e04-bf00-010606b936fb" width="300">

---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:
